interpreter/interpreter.go

```diff
@@ -0,0 +1,286 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The interpreter package provides functions to evaluate CEL programs against
+// a series of inputs and functions supplied at runtime.
+package interpreter
+
+import (
+	"fmt"
+	"github.com/golang/protobuf/proto"
+	"github.com/google/cel-go/interpreter/types"
+	"github.com/google/cel-go/interpreter/types/aspects"
+	"github.com/google/cel-go/interpreter/types/providers"
+	"github.com/google/cel-go/interpreter/functions"
+)
+
+// Interpreter generates a new Interpretable from a Program.
+type Interpreter interface {
+	// NewInterpretable returns an Interpretable from a Program.
+	NewInterpretable(program Program) Interpretable
+}
+
+// Interpretable can accept a given Activation and produce a result along with
+// an accompanying EvalState which can be used to inspect whether additional
+// data might be necessary to complete the evaluation.
+type Interpretable interface {
+	// Eval an Activation to produce an output and EvalState.
+	Eval(activation Activation) (interface{}, EvalState)
+}
+
+type exprInterpreter struct {
+	dispatcher   Dispatcher
+	typeProvider providers.TypeProvider
+}
+
+var _ Interpreter = &exprInterpreter{}
+
+// NewInterpreter builds an Interpreter from a Dispatcher and TypeProvider
+// which will be used throughout the Eval of all Interpretable instances
+// gerenated from it.
+func NewInterpreter(dispatcher Dispatcher, typeProvider providers.TypeProvider) *exprInterpreter {
+	return &exprInterpreter{dispatcher, typeProvider}
+}
+
+// StandardInterpreter builds a Dispatcher and TypeProvider with support
+// for all of the CEL builtins defined in the language definition.
+func StandardIntepreter(types ...proto.Message) *exprInterpreter {
+	dispatcher := NewDispatcher()
+	dispatcher.Add(functions.StandardBuiltins()...)
+	typeProvider := providers.NewTypeProvider(types...)
+	return NewInterpreter(dispatcher, typeProvider)
+}
+
+func (i *exprInterpreter) NewInterpretable(program Program) Interpretable {
+	// program needs to be pruned with the TypeProvider
+	return &exprInterpretable{
+		interpreter: i,
+		program:     program,
+		state:       NewEvalState()}
+}
+
+type exprInterpretable struct {
+	interpreter *exprInterpreter
+	program     Program
+	state       MutableEvalState
+}
+
+var _ Interpretable = &exprInterpretable{}
+
+func (i *exprInterpretable) Eval(activation Activation) (interface{}, EvalState) {
+	// register machine-like evaluation of the program with the given activation.
+	currActivation := activation
+	stepper := i.program.Instructions()
+	var resultId int64
+	for step, hasNext := stepper.Next(); hasNext; step, hasNext = stepper.Next() {
+		resultId = step.GetId()
+		switch step.(type) {
+		case *ConstExpr:
+			i.evalConst(step)
+		case *IdentExpr:
+			i.evalIdent(step, currActivation)
+		case *SelectExpr:
+			i.evalSelect(step, currActivation)
+		case *CallExpr:
+			i.evalCall(step, currActivation)
```

Comment:
Suggest doing the type assertion here where we can be sure it won't panic, rather than in the function.

```go
		case *CallExpr:
			i.evalCall(step.(*CallExpr), currActivation)
```

Reply:
SGTM, done.